graph/pull.go

```diff
@@ -256,6 +263,9 @@ func (s *TagStore) pullImage(r *registry.Session, out io.Writer, imgID, endpoint
 	// FIXME: Try to stream the images?
 	// FIXME: Launch the getRemoteImage() in goroutines
 
+	s.graph.Retain(history, r.ID())
+	defer s.graph.Release(history, r.ID())
+
```

Comment:
Oh, I see. Ignore my earlier comment about removing this section.

The `history` list will contain `imgID` which is already retained prior in `pullRepository()` and if all history is released again before this function returns, then the image ID will have been released prematurely, before `pullRepository()` has a chance to tag it. Changing this call to:

```go
defer s.graph.Release(history[1:], r.ID())
```
should release all but the first item in the list which is the already retained `imgID`.